### PR TITLE
feat(slack): Implement config UI

### PR DESCRIPTION
**Pull request by evanpurkhiser**
Here's a few screenshots

image

image

### CODE REVIEW

**Comment by getsentry-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">You should update CHANGES due to the size of this PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

**Comment by benvinegar** [comment]
@EvanPurkhiser did you need someone to look at this? It's been up for 2 weeks.

**Comment by EvanPurkhiser** [comment]
@benvinegar Yeah going to throw @MaxBittker and @billyvg  on this. I have a number of more commits locally that I'm organizing now to add onto this.

File: src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+    const {success, data} = message.data;
+
+    if (!success) {
+      // TODO: Error state here
+      return;
+    }
+
+    let itemList = [...this.state.itemList, data];
+    itemList = sortArray(itemList, i => i.name);
+    itemList = _.uniqBy(itemList, i => i.id);
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx]: In what case would we have duplicate ids here? Can we add a comment?

File: src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+    const {orgId} = this.props.params;
+
+    const options = {
+      method: 'DELETE',
+      success: () =>
+        this.setState({
+          itemList: this.state.itemList.filter(item => item.id !== integration.id),
+        }),
+
+      // TODO toast indoactor for errors / complete
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx]: Use `addSuccessMessage` and `addErrorMessage` action creators

File: src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+              <span className="icon icon-add" /> {t('Add Workspace')}
+            </Button>
+          </Box>
+        </Flex>
+      </PanelHeader>
+    );
+
+    const integrationList = integrations.map(integration => {
+      return (
+        <Row p={0} py={2} key={integration.id}>
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx]: Use `<PanelItem>` instead of `<Row>`

File: src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+            <IntegrationIcon src={integration.icon} />
+          </Box>
+          <Box px={2} flex={1}>
+            <IntegrationName>{integration.name}</IntegrationName>
+            <DomainName>{integration.domain_name}</DomainName>
+          </Box>
+          <Box mr={1} pr={2}>
+            <Confirm
+              message={t('Are you sure you want to remove this integration?')}
+              onConfirm={() => this.deleteIntegration(integration)}
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx]: Let's use `handle*` prefix for callback handlers

File: src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+    const titleIcon = <PluginIcon size={28} pluginId={provider.key} />;
+
+    const header = (
+      <PanelHeader disablePadding hasButtons>
+        <Flex align="center">
+          <Box px={2} flex="1">
+            {t('Workspaces')}
+          </Box>
+          <Box mr={1}>
+            <Button size="xsmall" onClick={() => this.launchAddIntegration(provider)}>
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx]: Let's use `handle*` prefix for callback handlers

File: src/sentry/static/sentry/app/views/projectPlugins/index.jsx [hunk 6]
```diff
@@ -27,15 +30,25 @@ class ProjectPluginsContainer extends React.Component {
   render() {
     let {loading, error, plugins} = this.props.plugins || {};
 
+    console.log(this.props);
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/projectPlugins/index.jsx]: console

File: src/sentry/static/sentry/app/views/projectPlugins/index.jsx [hunk 7]
```diff
@@ -27,15 +30,25 @@ class ProjectPluginsContainer extends React.Component {
...
+
+    // TODO: We shouldn't be passing all the props to OrgnizationIntegrations,
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/projectPlugins/index.jsx]: ^^

File: src/sentry/static/sentry/app/views/projectPlugins/organizationIntegrations.jsx [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    let {orgId} = this.props.params;
+
+    return [['config', `/organizations/${orgId}/config/integrations/`]];
+  }
+
+  renderBody() {
+    // TODO: This is for sure being passsed in to many things
+    let {orgId, projectId} = this.props.params;
+
+    console.log(this.props.params);
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/projectPlugins/organizationIntegrations.jsx]: ^^

File: src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+      );
+    });
+
+    return (
+      <React.Fragment>
+        <SettingsPageHeader title={provider.name} icon={titleIcon} />
+
+        <Panel>
+          {header}
+          <PanelBody>{integrationList}</PanelBody>
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx]: Does this need empty state?

File: src/sentry/static/sentry/app/views/projectPlugins/organizationIntegrations.jsx [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+    return (
+      <Panel>
+        <PanelHeader disablePadding={true}>
+          <Box px={2} flex="1">
+            {t('Global Integrations')}
+          </Box>
+        </PanelHeader>
+        <PanelBody>
+          {this.state.config.providers.map(provider => {
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/projectPlugins/organizationIntegrations.jsx]: empty state?

**Review by billyvg — COMMENTED** [review]
Left some comments, some screenshots would be nice

**Review by billyvg — COMMENTED** [review]
Left some comments, some screenshots would be nice

File: src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx (hunk 9 above)
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx]: e31f221

File: src/sentry/static/sentry/app/views/projectPlugins/organizationIntegrations.jsx (hunk 10 above)
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/views/projectPlugins/organizationIntegrations.jsx]: e31f221

File: src/sentry/static/sentry/app/routes.jsx [hunk 11]
```diff
@@ -340,6 +341,14 @@ const projectSettingsRoutes = [
     name="Integration Details"
     component={errorHandler(ProjectPluginDetails)}
   />,
+  // XXX(epurkhiser): This will live under project configurations for now until
+  // we've migrated enough integrations that it can live at the org level.
+  <Route
+    key="integrations/:providerKey/"
```
[the remark this answers is not in the document]
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/routes.jsx]: Rebased + 114d557d3ee80835a2ba622408fa2e13062f9371

File: src/sentry/static/sentry/app/routes.jsx [hunk 12]
```diff
@@ -340,6 +341,14 @@ const projectSettingsRoutes = [
...
+    path="integrations/:providerKey/"
+    name="Integration Configuration"
+    component={errorHandler(OrganizationIntegrationConfig)}
```
[the remark this answers is not in the document]
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/routes.jsx]: 114d557

File: src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx (hunk 2 above)
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx]: c08d1f0d85398d6cfc0f51cc9ee7890cc36d8485

File: src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx (hunk 3 above)
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx]: Comment is slightly wrong in commit but: 806805e

File: src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx (hunk 4 above)
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx]: 8e6cfe6

File: src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx (hunk 5 above)
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx]: 8e6cfe6

File: src/sentry/static/sentry/app/views/projectPlugins/index.jsx (hunk 6 above)
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/views/projectPlugins/index.jsx]: Removed in latest rebase.

File: src/sentry/static/sentry/app/views/projectPlugins/index.jsx (hunk 7 above)
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/views/projectPlugins/index.jsx]: fd28f12

File: src/sentry/static/sentry/app/views/projectPlugins/organizationIntegrations.jsx (hunk 8 above)
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/views/projectPlugins/organizationIntegrations.jsx]: removed.

File: src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx (hunk 9 above)
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx]: aa242db578afd1afd602e620bf65e8c11a18e531. Added above.

**Comment by billyvg** [comment]
JS looks good, there's a bunch of test failures though. What's the plan for release? Will this be behind a feature tag?

File: src/sentry/static/sentry/app/routes.jsx [hunk 13]
```diff
@@ -84,6 +84,7 @@ import ProjectUserReportSettings from './views/projectUserReportSettings';
 import ProjectUserReports from './views/projectUserReports';
 import ProjectPlugins from './views/projectPlugins';
 import ProjectPluginDetails from './views/projectPluginDetails';
+import OrganizationIntegrationConfig from './views/organizationIntegrationConfig';
```
Comment (billyvg) [inline src/sentry/static/sentry/app/routes.jsx]: Unused

**Comment by EvanPurkhiser** [comment]
Ready for one more once over @billyvg / cc @getsentry/workflow

File: src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx [hunk 14]
```diff
@@ -0,0 +1,305 @@
...
+const MetadataLink = styled.a`
+  display: block;
+  margin-bottom: 6px;
+`;
+
+const alertLinkMarkedRenderer = new marked.Renderer();
+alertLinkMarkedRenderer.paragraph = s => s;
+
+function computeCenteredWindow(width, height) {
+  const screenLeft = window.screenLeft != undefined ? window.screenLeft : screen.left;
```
Comment (MaxBittker) [inline src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx]: if this is copy pasted from somewhere, could you include a link in case someone is trying to debug it in the future?

File: src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx [hunk 15]
```diff
@@ -0,0 +1,305 @@
...
+    if (message.source !== this.dialog) {
+      return;
+    }
+
+    this.dialog = null;
+
+    const {success, data} = message.data;
+
+    if (!success) {
+      // TODO: Error state here
```
Comment (MaxBittker) [inline src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx]: when would this happen? at the very least capture this for raven? and/or add a failure indicator?

File: src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx [hunk 16]
```diff
@@ -0,0 +1,305 @@
...
+      return;
+    }
+
+    // Merge the new integration into the list. If we're updating an
+    // integration ovewrrite the old integration.
+    const keyedItems = _.keyBy(this.state.itemList, 'id');
+    const itemList = sortArray(
```
Comment (MaxBittker) [inline src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx]: wouldn't it be simpler to keep it as an object in state and then turn it to a sorted array in render only? to avoid going back and forth?

**Review by MaxBittker — COMMENTED** [review]
will finish review after lunch

File: src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx (hunk 14 above)
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx]: Copied from dcramer:

https://github.com/getsentry/sentry/blob/b0fa526b28df4a4b557c874298e9a2aa97ee3098/src/sentry/static/sentry/app/views/organizationIntegrations.jsx#L81-L97

(this whole other integration component will go away soon)

File: src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx [hunk 17]
```diff
@@ -0,0 +1,305 @@
...
+import Panel from './settings/components/panel';
+import PanelBody from './settings/components/panelBody';
+import PanelHeader from './settings/components/panelHeader';
+import PanelItem from './settings/components/panelItem';
+import PluginIcon from '../plugins/components/pluginIcon';
+import SettingsPageHeader from './settings/components/settingsPageHeader';
+import marked from '../utils/marked';
+
+const IntegrationIcon = styled.img`
+  height: 32px;
```
Comment (MaxBittker) [inline src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx]: not your fault, but some of these px values make me think we really need some abstraction for managing them

File: src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx [hunk 18]
```diff
@@ -0,0 +1,305 @@
...
+
+    return `${provider.name} Integration`;
+  }
+
+  getEndpoints() {
+    const {orgId} = this.props.params;
+
+    return [
+      ['config', `/organizations/${orgId}/config/integrations/`],
+      ['itemList', `/organizations/${orgId}/integrations/`, {query: {status: ''}}],
```
Comment (MaxBittker) [inline src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx]: why does it need the empty query? worth commenting?

File: src/sentry/static/sentry/app/views/projectPlugins/organizationIntegrations.jsx [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+
+const ProviderName = styled.div`
+  font-size: 16px;
+  font-weight: bold;
+  margin-bottom: 3px;
+`;
+
+const TeamName = styled.div`
+  color: ${p => p.theme.gray2};
+  font-size: 14px;
```
Comment (MaxBittker) [inline src/sentry/static/sentry/app/views/projectPlugins/organizationIntegrations.jsx]: yeah this stuff is way too manual, is it possible to pull it off theme?

Comment (MaxBittker) [reply src/sentry/static/sentry/app/views/projectPlugins/organizationIntegrations.jsx]: (not your fault)

**Review by MaxBittker — APPROVED** [review]
nice job, looks good to me % nits

File: src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx (hunk 16 above)
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx]: Nah, I'm ok with it working like this, Just have to have this merging logic since adding an integration is a create or update operation.

File: src/sentry/static/sentry/app/views/projectPlugins/organizationIntegrations.jsx (hunk 19 above)
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/views/projectPlugins/organizationIntegrations.jsx]: Probably more of a question for @ckj / @Chrissy / @billyvg

File: src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx (hunk 18 above)
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx]: Dunno, copied this from the other integrations page

File: src/sentry/static/sentry/app/views/projectPlugins/organizationIntegrations.jsx (hunk 19 above)
Comment (Chrissy) [reply src/sentry/static/sentry/app/views/projectPlugins/organizationIntegrations.jsx]: GREAT QUESTION! 

tldr; right now it's kind of anything goes, so do whatever.

In the future, we should probably have some kind of type hierarchy system in `theme.js`, which we can also use for sizing up elements and things. 

As for the avatar and padding and stuff, it's generally a good idea to write those values in some kind of relative unit (percentages, vws, ems) or to use a flex layout to achieve the look you want. That way, the component is more fluid and can be resized with a top-down approach rather than having to go in and mutate a bunch of pixel values on props. But this is all kind of later stuff and shouldn't block this pull.

File: tests/js/spec/views/projectPlugins/projectPlugins.spec.jsx [hunk 20]
```diff
@@ -7,16 +7,32 @@ describe('ProjectPlugins', function() {
...
+    projectId: project.slug,
+  };
 
   it('renders', function() {
-    wrapper = shallow(<ProjectPlugins params={params} plugins={plugins} />);
+    wrapper = shallow(
+      <ProjectPlugins
+        params={params}
+        plugins={plugins}
+        organization={{features: ['integrations-v3']}}
```
[the remark this answers is not in the document]
Comment (EvanPurkhiser) [reply tests/js/spec/views/projectPlugins/projectPlugins.spec.jsx]: This flag will be going away pretty soon.

File: src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx [hunk 21]
```diff
@@ -0,0 +1,305 @@
...
+            <Description
+              dangerouslySetInnerHTML={{__html: marked(provider.metadata.description)}}
+            />
+            <AuthorName>{t('By %s', provider.metadata.author)}</AuthorName>
+          </Box>
+          <Box ml={60}>
+            <MetadataLink href={provider.metadata.issue_url}>
+              Report an Issue
+            </MetadataLink>
+            <MetadataLink href={provider.metadata.source_url}>View Source</MetadataLink>
```
[the remark this answers is not in the document]
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx]: f97cc8d

File: src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx [hunk 22]
```diff
@@ -0,0 +1,305 @@
...
+        <Details>
+          <Box width={5 / 8}>
+            <Description
+              dangerouslySetInnerHTML={{__html: marked(provider.metadata.description)}}
+            />
+            <AuthorName>{t('By %s', provider.metadata.author)}</AuthorName>
+          </Box>
+          <Box ml={60}>
+            <MetadataLink href={provider.metadata.issue_url}>
+              Report an Issue
```
[the remark this answers is not in the document]
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx]: f97cc8d

File: src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx (hunk 15 above)
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx]: Looks like this doesn't happen at all now. I'm just going to add a failed indicator
